### DISCUSSION TITLE
pushConstants binding other data

### DISCUSSION THREAD

**Discussion by orchidshelley** [vsg-dev/VulkanSceneGraph#843] [General]
hi,
     I found that in vsg PushConstantRanges only binding with projection matrix and model viewMatrix.  I hope I can use PushConstantRange or PushConstants to bind with other data, such as changing color. Is there any way to do that?

Thanks a lot~

**Comment by vsg-dev** [comment]
When you set up your graphics pipelines you'll need to specify the [vsg::PushConstantRanges](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/state/PipelineLayout.h#L35) in the Pipeline setup, and then use assign the [vsg::PushConstants](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/state/PushConstants.h) to either a StateGroup above the subgaph you want to influence or added before the draw calls if you have a sequence of changes.

**Reply by orchidshelley** [reply]
I need more help. 

My pushconstants looks like such that:

> struct MyPushConstant{
     vsg::mat4 modelMatrix;
     vsg::vec4 baseColor;
     vsg::vec4 selectColor;
     vsg::vec4 otherData1;
      vsg::vec4 otherData2;
}

at the begining in the Pipeline setup， 

`vsg::PushConstantRanges pushConstantRanges{
        {VK_SHADER_STAGE_VERTEX_BIT, 0, 128} 
    };`

`auto pipelineLayout = vsg::PipelineLayout::create(vsg::DescriptorSetLayouts{ descriptorSetLayout }, pushConstantRanges);`

I think if I used such way ablove,  it is the the projection matrix and modelview matrix will bind with the pushConstants, not  my pushConstants.

I don't know when is the right time , where is the right place to push my constants.

is it in update time?But how to use it? I tried to use it in below way, but the code looks weird。

> void Mesh::update()
   {
       // m_sceneGraph type is ref_ptr<vsg::StateGroup>

      // Must I clear the stateCommands every frame? If I don't clear,  m_sceneGraph will add state command every frame.
       m_sceneGraph->stateCommands.clear();
       
      // Is there othe way to turn  vsg::mat4 to vsg::ref_ptr<vsg::Data>? I don't know,  so I used a  very roundabout way。
       auto matModelArray = vsg::mat4Array::create(1);
       matModelArray.set(0, myData.modelMatrix);
       auto matPC = vsg::PushConstants::create(VK_SHADER_STAGE_VERTEX_BIT, 0,  matModelArray.get());
       m_sceneGraph->add(matPC );

      //  the same as vsg::vec4,  I dont't know how to turn it to vsg::Data
      auto colorArray = vsg::vec4Array::create(1);
      colorArray .set(0, myData.baseColor);
      auto colorPC = vsg::PushConstants::create(VK_SHADER_STAGE_VERTEX_BIT,  64,  colorArray .get());
       m_sceneGraph->add(colorPC );
      
      // the two other vec4 are similar， just the offsets are different
     ...
  }

and the update is called in:

> while (m_viewer->advanceToNextFrame) {
      m_viewer->handleEvent();
      m_viewer->update();

     // call mesh update
     m_mesh->update();
    
     m_viewer->recordAndSubmit();
     m_viewer->present();
   }

if  I used in that way, my pushConstant will be covered with projection matrix and modelview matrix  in the m_viewer->recordAndSubmit().

I think I used something wrong.
What is the right way to use my own push constants?

**Comment by timoore** [comment]
Also, the minimum value of maxPushConstantsSize is only 128 bytes, which
VSG uses completely with the two matrices. I've read that on some mobile
hardware the actual limit is 64 bytes, with usage above that simulated in
some slow way.

Tim

**Reply by orchidshelley** [reply]
thank you

**Comment by vsg-dev** [comment]
PushConstants in Vulkan are a small block of memory that can be copied cheaply to the GPU as part of the command buffer.  You can't have multiple different PushConstants competing for the same space, so if you want extras above the projection and modelview matrices that the VSG uses for it's standard shaders and usage then you'll need to allocate after bigger than 128 bytes and place your additions after position 128, so the that VSG can set the projection at location 0 and modelview at location 64 without conflict.

As Tim mentions how much memory is available for push constants is limited by hardware/drivers so you need to query this.

As a general note, I would recommend against passing model matrices to shaders.  I know there are lots of shader examples doing this online but it's not a good practice as it leads to precision problems handling numerically large datasets.  The model matrix approach is fine for toy datasets but not anything beyond this.   

To avoid precision issues the VSG accumulates model matrices in the scene graph with the view matrix using double matrix math then passes to the GPU as a float modelview matrix.  This approach solve the precision issue so that models don't jitter around on screen instead stay solid on screen.  The use of modelview rather than model does maan that lighting and texgen calcs need to be done in eye space, so need to be passed to the GPU in eye space, again this is what the VSG does when passing light positions/direction - the eye space position is computed in eye space for you and passed to the GPU on each frame in eye space.
